File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int res = INT_MIN;
        dfs(root, res);
        return res;
    }
    int dfs(TreeNode* root, int &res){
        if( !root){
            return 0;
        }

        int maxLeft = dfs(root->left, res);
        int maxRight = dfs(root->right, res);
        maxLeft = max(maxLeft, 0);
        maxRight = max(maxRight, 0);

        res = max(res, root->val + maxLeft + maxRight);

        return root->val + max(maxLeft, maxRight);
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        self.res = 0
        max_val = root.val
        self.dfs(root, max_val)
        return self.res
    def dfs(self, root, max_val):
        if not root:
            return 0
        max_val = max(max_val, root.val)
        if root.val >= max_val:
            self.res+=1
        self.dfs(root.left, max_val)
        self.dfs(root.right, max_val)
        return 0